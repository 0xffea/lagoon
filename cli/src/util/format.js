// @flow

import R from 'ramda';
import {
  FORMAT,
  FORMAT_CHOICE_TABLE,
  FORMAT_CHOICE_SIMPLE,
  FORMAT_CHOICE_JSON,
  FORMAT_CHOICE_CSV,
} from '../config/globalOptions';
import { getConfig } from '../config';

import type { FormatData } from '../types/Format';

export default function format(
  data: FormatData,
  options: ?{
    // Optional transforms for data
    transforms?: { [key: string]: (FormatData) => FormatData },
  },
) {
  const config = getConfig();
  const configuredFormat = R.prop(FORMAT, config);
  const transformByFormat =
    R.path(['transforms', configuredFormat], options) ||
    // Default to just returning the data
    R.identity;
  const transformedData = transformByFormat(data);

  return R.cond([
    [
      R.equals(FORMAT_CHOICE_TABLE),
      R.always(require('../formatters/table').default(transformedData)),
    ],
    [
      R.equals(FORMAT_CHOICE_SIMPLE),
      R.always(require('../formatters/simple').default(transformedData)),
    ],
    [
      R.equals(FORMAT_CHOICE_JSON),
      R.always(require('../formatters/json').default(transformedData)),
    ],
    [
      R.equals(FORMAT_CHOICE_CSV),
      R.always(require('../formatters/csv').default(transformedData)),
    ],
  ])(configuredFormat);
}
