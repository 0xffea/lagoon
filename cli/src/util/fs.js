// @flow

/**
 * This module exists to promisify all the fs functionality
 */

import fs from 'fs';

import type { Stats } from 'fs';

export function readFile(filename: string, enc?: string): Promise<string> {
  return new Promise((resolve, reject) => {
    fs.readFile(filename, enc || 'utf8', (err, data) => {
      if (err) {
        reject(err);
      } else {
        resolve(data);
      }
    });
  });
}

export function writeFile(
  filename: string,
  data: Buffer | string,
  options?: Object | string,
): Promise<void> {
  return new Promise((resolve, reject) => {
    fs.writeFile(filename, data, options, (err) => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

export function deleteFile(filename: string): Promise<void> {
  return new Promise((resolve, reject) => {
    fs.unlink(filename, (err) => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

export function lstat(path: string): Promise<Stats> {
  return new Promise((resolve, reject) => {
    fs.lstat(path, (err, stat) => {
      if (err) {
        reject(err);
      } else {
        resolve(stat);
      }
    });
  });
}

export async function doesFileExist(file: string): Promise<boolean> {
  try {
    const stats = await lstat(file);

    if (stats.isFile()) {
      return true;
    }

    return false;
  } catch (err) {
    return false;
  }
}

export const statSync = fs.statSync;
