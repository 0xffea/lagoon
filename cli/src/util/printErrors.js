// @flow

import { red } from 'chalk';
import R from 'ramda';

type Cerr = typeof console.error;
type LaguError = { message: string };

export function printErrors(cerr: Cerr, ...errors: Array<LaguError>): number {
  R.compose(
    R.forEach(cerr),
    R.map(red),
    R.map(R.prop('message')),
  )(errors);
  return 1;
}

export function printNoConfigError(cerr: Cerr): number {
  return printErrors(cerr, {
    message:
      '.lagoon.yml config file not found. Please create one with "lagoon init".\nOnline documentation: https://github.com/amazeeio/lagoon/blob/master/cli/README.md#lagoon-init',
  });
}

export function printProjectConfigurationError(cerr: Cerr): number {
  return printErrors(cerr, {
    message:
      'No project configured. Please create a .lagoon.yml config file with "lagoon init" or pass a project to this command via the --project option.\nOnline documentation: https://github.com/amazeeio/lagoon/blob/master/cli/README.md#lagoon-init',
  });
}

export function printGraphQLErrors(
  cerr: Cerr,
  ...errors: Array<LaguError>
): number {
  if (
    R.find(R.propEq('message', 'Unauthorized - Bearer Token Required'))(errors)
  ) {
    return printErrors(cerr, {
      message:
        'No authentication token found. Please log in first with "lagoon login".\nOnline documentation: https://github.com/amazeeio/lagoon/blob/master/cli/README.md#lagoon-login',
    });
  }
  const errorMessage =
    R.length(errors) === 1
      ? 'Oops! The lagoon API returned an error:'
      : 'Oops! The lagoon API returned errors:';
  return printErrors(cerr, { message: errorMessage }, ...errors);
}
