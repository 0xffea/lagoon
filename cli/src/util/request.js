// @flow

import defer from './defer';

export type RequestOptions = {
  hostname: string,
  path?: string,
  port?: number | string,
  method: 'POST' | 'GET',
  headers?: Object,
  rejectUnauthorized?: boolean,
  body?: string,
};

export default function request(options: RequestOptions): Promise<Object> {
  const { body, port = 443 } = options;

  const def = defer();

  const thing = Number(port) === 80 ? require('http') : require('https');

  const req = thing.request(options, res => {
    res.setEncoding('utf8');

    let rawData = '';
    res.on('data', chunk => {
      rawData += chunk;
    });
    res.on('end', () => {
      try {
        let parsed = JSON.parse(rawData);
        def.resolve(parsed);
      }
      catch (e) {
        def.reject(e);
      }
    });
  });

  if (body) {
    req.write(body);
  }

  req.end();

  req.on('error', e => def.reject(e));

  return def.promise;
}
