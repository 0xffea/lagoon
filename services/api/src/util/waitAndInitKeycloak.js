// @flow

const KeycloakAdminClient = require('keycloak-admin').default;
const logger = require('../logger');

/* ::

type ConnectionSettings = {
  baseUrl?: string,
  realmName?: string,
  requestConfig?: Object,
}

type UserSettings = {
  username?: string,
  password?: string,
  grantType?: string,
  clientId?: string,
};

*/

async function waitAndInitKeycloak(
  connectionSettings /* : ConnectionSettings */,
  userSettings /* : UserSettings */,
) {
  let keycloakClient;
  let keycloakReady = false;

  keycloakClient = new KeycloakAdminClient(connectionSettings);

  do {
    try {
      await keycloakClient.auth(userSettings);
      if (!(await keycloakClient.realms.findOne({ realm: 'lagoon' }))) {
        throw new Error('The "lagoon" realm has not been created yet.')
      }

      keycloakClient.setConfig({ realmName: 'lagoon' });
      keycloakReady = true;
    } catch (err) {
      logger.debug(`Waiting for Keycloak to start... (error was ${err})`);
      await new Promise(resolve => setTimeout(resolve, 2000));
    }

  } while (!keycloakReady);

  setInterval(async () => await keycloakClient.auth(userSettings), 55*1000);

  if (!keycloakClient) {
    throw new Error('Keycloak client not initialized!');
  }

  logger.debug('Connected to Keycloak');

  return keycloakClient;
}

module.exports = waitAndInitKeycloak;
