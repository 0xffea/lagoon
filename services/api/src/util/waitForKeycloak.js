// @flow

const KeycloakAdminClient = require('keycloak-admin').default;
const logger = require('../logger');

/* ::

type ConnectionSettings = {
  baseUrl?: string,
  realmName?: string,
  requestConfig?: Object,
}

type UserSettings = {
  username?: string,
  password?: string,
  grantType?: string,
  clientId?: string,
};

*/

async function waitForKeycloak(
  connectionSettings /* : ConnectionSettings */,
  userSettings /* : UserSettings */,
) {
  let keycloakClient;
  let keycloakReady = false;

  do {
    try {
      keycloakClient = new KeycloakAdminClient(connectionSettings);
      await keycloakClient.auth(userSettings);
      keycloakReady = true;
    } catch (err) {
      logger.debug('Waiting for Keycloak to start...');
      await new Promise(resolve => setTimeout(resolve, 2000));
    }
  } while (!keycloakReady);

  if (!keycloakClient) {
    throw new Error('Keycloak client not initialized!');
  }

  return keycloakClient;
}

module.exports = waitForKeycloak;
