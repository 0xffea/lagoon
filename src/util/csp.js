// @flow

import { takeAsync } from 'js-csp';
import { fromColl } from 'js-csp/lib/csp.operations';

import typeof { Channel } from 'js-csp/es/impl/channels';

// TODO: Make Channel generic, so we at least tell which resolve type Promise yields
export function resolveChannel(channel: Channel, keepOpen?: boolean = false): Promise<mixed|Error> {
  return new Promise((res, rej) => {
    takeAsync(channel, (val) => { 
      res(val);

      if (keepOpen) {
        channel.close();
      }
    });
  });
}

export function maybeResolveChannel(channel: Channel, keepOpen?: boolean = false): Promise<mixed> {
  return new Promise((res, rej) => {
    takeAsync(channel, (val) => { 
      if (val instanceof Error) {
        rej(val);
      }
      else {
        res(val);
      }

      if (keepOpen) {
        channel.close();
      }
    });
  });
}


export function fail(msg: string): Channel {
  return fromColl([new Error(msg)]);
}
