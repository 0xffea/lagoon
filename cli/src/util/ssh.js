// @flow

import inquirer from 'inquirer';
import R from 'ramda';
import { Client } from 'ssh2';
import { fileExists } from '../util/fs';
import { printErrors } from '../printErrors';

type ExecOptions = {};

type Connection = {
  exec: (
    command: string,
    options?: ExecOptions,
    callback: Function,
  ) => Connection,
  on: (event: string, callback: Function) => Connection,
  end: () => Connection,
};

type ConnectArgs = {
  host: string,
  port: number,
  username: string,
  privateKey: string | Buffer,
  passphrase?: string,
};

export async function sshConnect(args: ConnectArgs): Promise<Connection> {
  return new Promise((resolve, reject) => {
    const connection = new Client();

    connection.on('ready', () => {
      resolve(connection);
    });

    connection.on('error', (error) => {
      reject(error);
    });

    connection.connect(args);
  });
}

export async function sshExec(
  connection: Connection,
  command: string,
  options?: ExecOptions = {},
): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    connection.exec(command, options, (error, stream) => {
      if (error) {
        reject(error);
      } else {
        stream.on('data', (data) => {
          resolve(data);
        });

        stream.stderr.on('data', (data) => {
          reject(new Error(data));
        });
      }
    });
  });
}

type GetPrivateKeyPathArgs = {
  fileExistsAtDefaultPath: boolean,
  defaultPrivateKeyPath: string,
  identityOption?: string,
  cerr: typeof console.error,
};

export const getPrivateKeyPath = async (
  args: GetPrivateKeyPathArgs,
): Promise<string> =>
  R.cond([
    // If the identity option for the command has been specified and the file at the path exists, use the value of that
    [
      R.propSatisfies(
        // Option is not null or undefined
        R.complement(R.isNil),
        'identityOption',
      ),
      R.prop('identityOption'),
    ],
    // If a file exists at the default private key path, use that
    [R.prop('fileExistsAtDefaultPath'), R.prop('defaultPrivateKeyPath')],
    // If none of the previous conditions have been satisfied, ask the user if they want to overwrite the file
    [
      R.T,
      async ({ cerr }) => {
        // TODO: Do now.
        // TODO: Ramdaify / FPify?
        let privateKeyPath = '';
        let privateKeyExists = false;
        while (!privateKeyPath || !privateKeyExists) {
          // eslint-disable-next-line no-await-in-loop
          const promptReturn = await inquirer.prompt([
            {
              type: 'input',
              name: 'privateKeyPath',
              message: 'Path to private key file',
            },
          ]);
          privateKeyPath = promptReturn.privateKeyPath;

          // eslint-disable-next-line no-await-in-loop
          privateKeyExists = await fileExists(privateKeyPath);
          if (!privateKeyExists) {
            printErrors(cerr, 'File does not exist at given path!');
          }
        }
        return privateKeyPath;
      },
    ],
  ])(args);

export const getPrivateKeyPassphrase = async (
  privateKeyHasEncryption: boolean,
): Promise<string> =>
  R.ifElse(
    // If the private key doesn't have encryption...
    R.not,
    // ... return an empty string
    R.always(''),
    // If the private key has encryption, ask for the password
    async () => {
      const { passphrase } = await inquirer.prompt([
        {
          type: 'password',
          name: 'passphrase',
          message: 'Private key password (never saved)',
          default: '',
        },
      ]);
      return passphrase;
    },
  )(privateKeyHasEncryption);
