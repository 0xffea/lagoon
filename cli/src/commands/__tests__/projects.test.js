// @flow

import { queryGraphQL } from '../../util/queryGraphQL';
import { listProjects } from '../projects';

jest.mock('../../util/queryGraphQL');

const _mock = (mockFn: any): JestMockFn<any, any> => mockFn;

describe('listProjects', () => {
  it('should list details for multiple projects', async () => {
    _mock(queryGraphQL).mockImplementationOnce(() =>
      Promise.resolve({
        data: {
          allProjects: [
            {
              name: 'credentialstest-project1',
              git_url: 'ssh://git@192.168.99.1:2222/git/project1.git',
              branches: 'true',
              pullrequests: null,
              created: '2018-01-15 11:09:35',
            },
            {
              name: 'credentialstest-project2',
              git_url: 'ssh://git@192.168.99.1:2222/git/project2.git',
              branches: 'true',
              pullrequests: null,
              created: '2018-01-15 11:09:35',
            },
          ],
        },
      }),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await listProjects({
      clog,
      cerr,
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should print notice on empty projects array', async () => {
    _mock(queryGraphQL).mockImplementationOnce(() => Promise.resolve([]));

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await listProjects({
      clog,
      cerr,
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should display error, if GraphQL sends error messages', async () => {
    _mock(queryGraphQL).mockImplementationOnce(() =>
      Promise.resolve({
        errors: [{ message: 'something something error' }],
      }),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await listProjects({
      clog,
      cerr,
    });

    expect(code).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });
});
