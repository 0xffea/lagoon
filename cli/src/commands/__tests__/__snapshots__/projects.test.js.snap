// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handler should display error, if GraphQL sends error messages 1`] = `
Array [
  Array [
    "[31mOops! The Lagoon API returned an error:[39m",
  ],
  Array [
    "[31msomething something error[39m",
  ],
]
`;

exports[`handler should list details for multiple projects 1`] = `
Array [
  Array [
    "╔══════════════════════════╤══════════════════════════════════════════════╤══════════╤═══════════════╤═════════════════════╗
║ Project                  │ Git URL                                      │ Branches │ Pull Requests │ Created             ║
╟──────────────────────────┼──────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ credentialstest-project1 │ ssh://git@192.168.99.1:2222/git/project1.git │ true     │ null          │ 2018-01-15 11:09:35 ║
╟──────────────────────────┼──────────────────────────────────────────────┼──────────┼───────────────┼─────────────────────╢
║ credentialstest-project2 │ ssh://git@192.168.99.1:2222/git/project2.git │ true     │ null          │ 2018-01-15 11:09:35 ║
╚══════════════════════════╧══════════════════════════════════════════════╧══════════╧═══════════════╧═════════════════════╝
",
  ],
]
`;

exports[`handler should print notice on empty projects array 1`] = `
Array [
  Array [
    "No projects found.",
  ],
]
`;
